File: JJUsers/CLayoutDir/OSTree.cpp
//-*-C++-*-
//*****************************************************************************
//*****************************************************************************
#ifndef OS_TREE_CPP
#define OS_TREE_CPP
//*****************************************************************************
#include "OSTree.h"
//*****************************************************************************
template <typename T>
OSNode<T>::OSNode (T data, Color color) : data(data), color(color) {}

template <typename T>
OSNode<T>::~OSNode () {
  if (left != NULL) 
    delete left;
  if (right != NULL)
    delete right;
}

template <typename T>
T OSNode<T>::getData () const { return data; }

template <typename T>
size_t OSNode<T>::getSize () const { 
  return size;
}

template <typename T>
void OSNode<T>::setSize (size_t newSize) {
  size = newSize;
}

template <typename T>
void OSNode<T>::updateSize () {
  size = 1;
  if (left != NULL)
    size += left->getSize();
  if (right != NULL) 
    size += right->getSize();
}
//*****************************************************************************

// Constructors
template <typename T, typename C>
OSTree<T,C>::OSTree () : root(NULL) {}

// Destructors
template <typename T, typename C>
OSTree<T,C>::~OSTree () { 
  if (root != NULL) 
    delete root; 
  root = NULL;
}

// Accessor Methods
template <typename T, typename C>
Color OSTree<T,C>::colorOf (OSNode<T> *node) const {
  if (node == NULL)
    return BLACK;
  return node->color;
}

template <typename T, typename C>
size_t OSTree<T,C>::sizeOf (OSNode<T> *node) const {
  if (node == NULL)
    return 0;
  return node->size;
}

template <typename T, typename C>
void OSTree<T,C>::setColor (OSNode<T> * node, Color color) {
  assert(node != NULL || color == BLACK);
  if (node == NULL)
    return;
  node->color = color;
}

//template <typename T, typename C>
//OSNode<T> * OSTree<T,C>::getPredecessorOf (OSNode<T> *) const {}


template <typename T, typename C>
OSNode<T> * OSTree<T,C>::getSuccessorOf (OSNode<T> *x) const {
  assert(x != NULL);
  OSNode<T> *y;
  
  if ((y = x->right) != NULL) { // assignment to y is intentional 
  while(y->left != NULL) { // returns the minium of the right subtree of x 
      y = y->left;
    }
    return y;
  } 
  y = x->parent;
  while(y != NULL && x == y->right) { 
    x = y;
    y = y->parent;
  }
  return y;
}

template <typename T, typename C>
void OSTree<T,C>::showTree () {
  showTree(root);
}

template <typename T, typename C>
void OSTree<T,C>::showTree (OSNode<T> *node) {
  if (node != NULL) {
    showTree(node->left);
    cout << *node << endl;
    showTree(node->right);
  }
}


//template <typename T, typename C>
//OSNode<T> * OSTree<T,C>::search (T t) {}

//template <typename T, typename C>
//void OSTree<T,C>::checkAssumptions () {}


// Mutator Methods
template <typename T, typename C>
OSNode<T> * OSTree<T,C>::insert (T node) {
  //  assert(node != NULL);
  OSNode<T> *y;
  OSNode<T> *x;
  OSNode<T> *newNode = new OSNode<T>(node, RED);
  x = newNode;
  insertHelp(x);
  
  while (x != root && colorOf(x->parent) == RED) {
    if (x->parent == x->parent->parent->left) {
      y = x->parent->parent->right;
      if (colorOf(y) == RED) {

        /*
         * case 1: x's uncle is red: recolor
         *          (Cb)                     x(Cr)     
         *         /   \     	              /   \     
         *      (Br)     (Dr)	           (Bb)     (Db)
         *      / \     / \     ===>      / \     / \  
         *   x(Ar) 'c  'd  'e	         (Ar) 'c  'd  'e
         *    / \            	         / \            
         *  'a  'b           	       'a  'b           
         */
        
        setColor(x->parent, BLACK);
        setColor(y, BLACK);
        setColor(x->parent->parent, RED);
        x = x->parent->parent;
      } else {
        if (x == x->parent->right) {
          // case 2: x is a right child: rotate to get to case 3
          x = x->parent;
          leftRotate(x);
          // now x is a left child
        }
        // case 3: x is a left child: recolor and rotate grandparent
        setColor(x->parent, BLACK);
        setColor(x->parent->parent, RED);
        rightRotate(x->parent->parent);
      }
    } else {
      // Right case: x's parent is a right child	
      // symmetric with left case
      y = x->parent->parent->left;
      if (colorOf(y) == RED) {
        // case 1: x's uncle is red: recolor
        setColor(x->parent, BLACK);
        setColor(y,BLACK);
        setColor(x->parent->parent, RED);
        x = x->parent->parent;
      } else {
        if (x == x->parent->left) {
	  // case 2: x is a left child: rotate to get to case 3
          x = x->parent;
          rightRotate(x);
          // now x is a right child
        }
        // case 3: x is a right child: recolor and rotate grandparent
        setColor(x->parent, BLACK);
        setColor(x->parent->parent, RED);
        leftRotate(x->parent->parent);
      }
    }
  }
  setColor(root, BLACK);
  return newNode;
}

template <typename T, typename C>
void OSTree<T,C>::remove (OSNode<T> * z) {
  assert(z != NULL);
  OSNode<T> *x;
  OSNode<T> *y;
  OSNode<T> *w;

  if (z->left == NULL || z->right == NULL) {
    y = z;
  } else {
    y = getSuccessorOf(z);
  }
  if (y->left != NULL) {
    x = y->left;    
  } else {
    x = y->right;
  }
  if (x != NULL) {
    x->parent = y->parent;
  }
  if (y->parent == NULL) {
    root = x;
  } else {
    if (y == y->parent->left) {
      y->parent->left = x;
    } else {
      y->parent->right = x;
    }
    w = y->parent;
    while (w != NULL) {
      w->setSize(w->getSize() - 1);
      w = w->parent;
    }
  }
  if (y != z) {
    // copy y fields to z
    z->data = y->data;
    z->size = y->size;
    setColor(z, colorOf(y));
  }
  if (colorOf(y) == BLACK) {
    removeHelp(x);
  }
  delete y;
}

// Helper Methods
template <typename T, typename C>
void OSTree<T,C>::leftRotate (OSNode<T> *x) {
  assert(x != NULL);
  OSNode<T> *y;
  
  y = x->right;
  x->right = y->left;
  if (y->left != NULL) {
    y->left->parent = x;
  }
  y->parent = x->parent;
  if (x->parent == NULL) {
    root = y;
  } else {
    if (x == x->parent->left) {
      x->parent->left = y;
    } else {
      x->parent->right = y;
    }
  }
  y->left = x;
  x->parent = y;
  y->setSize(x->getSize());
  x->updateSize();
}

template <typename T, typename C>
void OSTree<T,C>::rightRotate (OSNode<T> *y) {
  assert(y != NULL);
  OSNode<T> *x;
  x = y->left;
  y->left = x->right;
  if (x->right != NULL) {
    x->right->parent = y;
  }
  x->parent = y->parent;
  if (y->parent == NULL) {
    root = x;
  } else {
    if (y == y->parent->left) {
      y->parent->left = x;
    } else {
      y->parent->right = x;
    }
  }
  x->right = y;
  y->parent = x;
  x->setSize(y->getSize());
  y->updateSize();
}

template <typename T, typename C>
void OSTree<T,C>::insertHelp (OSNode<T> *z) {
  assert(z != NULL);
  OSNode<T> *x;
  OSNode<T> *y;
  
  z->left = z->right = NULL;

  y = NULL;
  x = root;
  while (x != NULL) {
    x->setSize(x->getSize() + 1);
    y = x;
    if (C()(z->getData(), x->getData())) {
      x = x->left;
    } else {
      x = x->right;
    }
  }
  z->parent = y;
  if (y == NULL) {
    root = z;
  } else {
    if (C()(z->getData(), y->getData())) {
      y->left = z;     
    } else {
      y->right = z;
    }
  }
  z->setSize(1);
}

template <typename T, typename C>
void OSTree<T,C>::removeHelp (OSNode<T> *x, OSNode<T> *p) {
  assert(p != NULL);
  OSNode<T> *w;
  while (x != root && colorOf(x) == BLACK) {
    if (x == p->left) {
      w = p->right;
      if (colorOf(w) == RED) {
        setColor(w, BLACK);
        setColor(p, RED);
        rotateLeft(p);
        w = p->right;
      }
      if (colorOf(w->left) == BLACK && colorOf(w->right == BLACK)) {
        setColor(w, RED);
        x = p;
        p = p->parent;
      } else {
        if (colorOf(w->right) == BLACK) {
          setColor(w->left, BLACK);
          setColor(w, RED);
          rotateRight(w);
          w = p->right;
        }
        setColor(w, colorOf(p));
        setColor(p, BLACK);
        setColor(w->right, BLACK);
        rotateLeft(p);
        x = root;
      }
    } else {
      w = p->left;
      if (colorOf(w) == RED) {
        setColor(w, BLACK);
        setColor(p, RED);
        rotateRight(p);
        w = p->left;
      }
      if (colorOf(w->right) == BLACK && colorOf(w->left == BLACK)) {
        setColor(w, RED);
        x = p;
        p = p->parent;
      } else {
        if (colorOf(w->left) == BLACK) {
          setColor(w->right, BLACK);
          setColor(w, RED);
          rotateLeft(w);
          w = p->left;
        }
        setColor(w, colorOf(p));
        setColor(p, BLACK);
        setColor(w->left, BLACK);
        rotateRight(p);
        x = root;
      }      
    }
  }
  setColor(x, BLACK);
}

template <typename T, typename C>
OSNode<T> * OSTree<T,C>:: select (size_t i) const {
  return select(root, i);
}

template <typename T, typename C>
OSNode<T> * OSTree<T,C>:: select (OSNode<T> *x, size_t i) const {
  assert(x != NULL);
  size_t r = sizeOf(x->left) + 1;
  if (i == r) {
    return x;
  } else if (i < r) {
    return select(x->left, i);
  } else {
    return select(x->right, i - r);
  }
}

template <typename T, typename C>
size_t OSTree<T,C>::rank (OSNode<T> *x) const {
  assert(x != NULL);
  size_t r = sizeOf(x->left) + 1;
  OSNode<T> *y = x;
  while (y != root) {
    if (y == y->parent->right)
      r += sizeOf(y->parent->left) + 1;
    y = y->parent;
  }
  return r;
}
//*****************************************************************************
#endif
//*****************************************************************************
//*****************************************************************************

File: Archive/IDVDir/Source/OSTree.java
//*****************************************************************************
//*****************************************************************************
/* 
   Steven Trac
   OSTree - Ordered Statistics Tree
*/
//*****************************************************************************
import java.util.*;
import java.io.*;
import javax.swing.*;
import java.awt.*;
//*****************************************************************************
public class OSTree<E> {
  private Comparator<E> compare;

  private enum RBColor { RED, BLACK };
  public class Node {
    E data;
    RBColor color;
    int size;

    Node left;
    Node right;
    Node parent;

    Node (E data, RBColor color) {
      this.data = data;
      this.color = color;
      left = null;
      right = null;
      parent = null;
    }      
    void show () {
      System.out.println(data);
    }
    void updateSize () {
      size = 1;
      if (left != null)
        size += left.size;
      if (right != null)
        size += right.size;
    }
  }

  private Node root;

  public OSTree (Comparator<E> compare) {
    this.compare = compare;
    root = null;
  }


  private void leftRotate (Node x) {
    Node y;
    
    y = x.right;
    x.right = y.left;
    if (y.left != null) {
      y.left.parent = x;
    }
    y.parent = x.parent;
    if (x.parent == null) {
      root = y;
    } else {
      if (x == x.parent.left) {
        x.parent.left = y;
      } else {
        x.parent.right = y;
      }
    }
    y.left = x;
    x.parent = y;
    y.size = x.size;
    x.updateSize();
  }
  private void rightRotate (Node y) {
    Node x;

    x = y.left;
    y.left = x.right;
    if (x.right != null) {
      x.right.parent = y;
    }
    x.parent = y.parent;
    if (y.parent == null) {
      root = x;
    } else {
      if (y == y.parent.left) {
      y.parent.left = x;
      } else {
        y.parent.right = x;
      }
    }
    x.right = y;
    y.parent = x;
    x.size = y.size;
    y.updateSize();
  }
  private void insertHelp (Node z) {
    Node x;
    Node y;
    
    z.left = z.right = null;
    
    y = null;
    x = root;
    while (x != null) {
      x.size = x.size + 1;
      y = x;
      if (compare.compare(z.data, x.data) == -1) {
        x = x.left;
      } else {
        x = x.right;
      }
    }
    z.parent = y;
    if (y == null) {
      root = z;
    } else {
      if (compare.compare(z.data, y.data) == -1) {
        y.left = z;     
      } else {
        y.right = z;
      }
    }
    z.size = 1;
  }
  private int sizeOf (Node node) {
    if (node == null)
      return 0;
    return node.size;
  }
  private RBColor colorOf (Node node) {
    if (node == null) 
      return RBColor.BLACK;
    return node.color;
  }
  private void setColor (Node node, RBColor color) {
    if (node == null)
      return;
    node.color = color;
  }
  
  //private void removeHelp (Node x) {}    
  //public void remove (E data) {}
  //private Node getPredecessorOf (Node x) {}
  //private Node getSuccessorOf (Node x) {}
  //private Node find (E data, Node x) {}
  public Node insert (E data) {
    Node y;
    Node x;
    Node newNode = new Node(data, RBColor.RED);
    x = newNode;
    insertHelp(x);
    
    while (x != root && colorOf(x.parent) == RBColor.RED) {
      if (x.parent == x.parent.parent.left) {
        y = x.parent.parent.right;
        if (colorOf(y) == RBColor.RED) {
          
          /*
           * case 1: x's uncle is red: recolor
           *          (Cb)                     x(Cr)     
           *         /   \     	              /   \     
           *      (Br)     (Dr)	           (Bb)     (Db)
           *      / \     / \     ===>      / \     / \  
           *   x(Ar) 'c  'd  'e	         (Ar) 'c  'd  'e
           *    / \            	         / \            
           *  'a  'b           	       'a  'b           
           */          
          setColor(x.parent, RBColor.BLACK);
          setColor(y, RBColor.BLACK);
          setColor(x.parent.parent, RBColor.RED);
          x = x.parent.parent;
        } else {
          if (x == x.parent.right) {
            // case 2: x is a right child: rotate to get to case 3
            x = x.parent;
            leftRotate(x);
            // now x is a left child
          }
          // case 3: x is a left child: recolor and rotate grandparent
          setColor(x.parent, RBColor.BLACK);
          setColor(x.parent.parent, RBColor.RED);
          rightRotate(x.parent.parent);
        }
      } else {
      // Right case: x's parent is a right child	
      // symmetric with left case
        y = x.parent.parent.left;
        if (colorOf(y) == RBColor.RED) {
          // case 1: x's uncle is red: recolor
          setColor(x.parent, RBColor.BLACK);
          setColor(y,RBColor.BLACK);
          setColor(x.parent.parent, RBColor.RED);
          x = x.parent.parent;
        } else {
          if (x == x.parent.left) {
            // case 2: x is a left child: rotate to get to case 3
            x = x.parent;
            rightRotate(x);
            // now x is a right child
          }
          // case 3: x is a right child: recolor and rotate grandparent
          setColor(x.parent, RBColor.BLACK);
          setColor(x.parent.parent, RBColor.RED);
          leftRotate(x.parent.parent);
        }
      }
    }
    setColor(root, RBColor.BLACK);
    return newNode;    
  }
  private void showTree (Node node) {
    if (node != null) {
      showTree(node.left);
      node.show();
      showTree(node.right);
    }
  }
  public void showTree () {
    showTree(root);
  }

  public Node select (int i) {
    return select(root, i);
  }
  public Node select (Node x, int i) {
    int r = sizeOf(x.left) + 1;
    if (i == r) {
      return x;
    } else if (i < r) {
      return select(x.left, i);
    } else {
      return select(x.right, i - r);
    }
  }
  int rank (Node x) {
    int r = sizeOf(x.left) + 1;
    Node y = x;
    while (y != root) {
      if (y == y.parent.right)
        r += sizeOf(y.parent.left) + 1;
      y = y.parent;
    }
    return r;
  }
  
}
